### PR TITLE
Fix: Replaces Tavis.UriTemplates lib. with local project implementation

### LINKED ISSUE #331: Bug in Resolving Permissions Request URLs Correctly in DevX API
The `Tavis.UriTemplates` library that DevX API uses to resolve permission request urls when fetching permissions data has a bug.  

Executing this uri: https://graphexplorerapi.azurewebsites.net/permissions?requesturl=/organization/{id}/settings/iteminsights&method=GET&scopeType=DelegatedWork fetches scopes for this request url: "/groups/{id}/settings/{id}", returning the below permissions:

```
[
  {
    "value":"Directory.Read.All",
    "consentDisplayName":"Read directory data",
    "consentDescription":"Allows the app to read data in your organization's directory.",
    "isAdmin":true
  },
  {
    "value":"Directory.ReadWrite.All",
    "consentDisplayName":"Read and write directory data",
    "consentDescription":"Allows the app to read and write data in your organization's directory, such as other users, groups.  It does not allow the app to delete users or groups, or reset user passwords.",
    "isAdmin":true},
  {
    "value":"Directory.AccessAsUser.All",
    "consentDisplayName":"Access the directory as you",
    "consentDescription":"Allows the app to have the same access to information in your work or school directory as you do.",
    "isAdmin":true
    }
  ]
```
... instead of the intended permissions for the request url: "/organization/{id}/settings/iteminsights"

```
[
  {
    "value":"User.Read",
    "consentDisplayName":"Sign you in and read your profile",
    "consentDescription":"Allows you to sign in to the app with your organizational account and let the app read your profile. It also allows the app to read basic company information.",
    "isAdmin":false},
  {
    "value":"User.ReadBasic.All",
    "consentDisplayName":"Read all users' basic profiles",
    "consentDescription":"Allows the app to read a basic set of profile properties of other users in your organization on your behalf. Includes display name, first and last name, email address and photo.",
    "isAdmin":false
  },
  {
    "value":"User.Read.All",
    "consentDisplayName":"Read all users' full profiles",
    "consentDescription":"Allows the app to read the full set of profile properties, reports, and managers of other users in your organization, on your behalf.",
    "isAdmin":true
    
  },
  {
    "value":"Directory.Read.All",
    "consentDisplayName":"Read directory data",
    "consentDescription":"Allows the app to read data in your organization's directory.",
    "isAdmin":true
  },
  {
    "value":"User.ReadWrite",
    "consentDisplayName":"Read and update your profile",
    "consentDescription":"Allows the app to read your profile, and discover your group membership, reports and manager. It also allows the app to update your profile information on your behalf.",
    "isAdmin":false
  },
  {
    "value":"User.ReadWrite.All",
    "consentDisplayName":"Read and write all users' full profiles",
    "consentDescription":"Allows the app to read and write the full set of profile properties, reports, and managers of other users in your organization, on your behalf.",
    "isAdmin":true
  },
  {
    "value":"Directory.ReadWrite.All",
    "consentDisplayName":"Read and write directory data",
    "consentDescription":"Allows the app to read and write data in your organization's directory, such as other users, groups.  It does not allow the app to delete users or groups, or reset user passwords.",
    "isAdmin":true
  },
  {
    "value":"Directory.AccessAsUser.All",
    "consentDisplayName":"Access the directory as you",
    "consentDescription":"Allows the app to have the same access to information in your work or school directory as you do.",
    "isAdmin":true
    }
  ]
```

 
[AB#6333](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/6333)

### PR DESCRIPTION
**Pull request by irvinesunday**
Fixes #331 

**Proposes**:
- Replacing the `Tavis.UriTemplates` library with a local project implementation of the core functionalities from the former. Code has been ported from the [UriTemplateTable.cs](https://github.com/tavis-software/Tavis.UriTemplates/blob/master/src/UriTemplates/UriTemplateTable.cs), [UriTemplate.cs](https://github.com/tavis-software/Tavis.UriTemplates/blob/master/src/UriTemplates/UriTemplate.cs) and [UriTemplateTests.cs](https://github.com/tavis-software/Tavis.UriTemplates/blob/master/src/UriTemplateTests/UriTemplateTableTests.cs) classes. Some code refactoring and renaming have been applied.
The local project implementation is used to match request urls for MSGraph permissions and OpenAPI paths against pre-defined uri templates. The previous library dependency offered similar functionality, however, the pattern matching wasn't strictly matching the whole path, leading to a breaking experience for Graph Explorer, as highlighted in the referenced issue #332 . The fix, as described in this other [PR](https://github.com/tavis-software/Tavis.UriTemplates/pull/68) included adding a regex lookbehind to ensure strict matching.
The decision to port the functional code from the original public repo. to this solution was influenced by DevX API needing an urgent fix for the above indicated issue, but which would likely cause breaking experiences for other scenarios for other customers of the `Tavis.UriTemplates` lib.
- Additionally, failing tests from the `OpenAPIService.Test` class have been updated.

[AB#6124](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/6124)

### CODE REVIEW

File: OpenAPIService.Test/OpenAPIServiceShould.cs [hunk 1]
```diff
@@ -297,7 +297,7 @@ public void RemoveRootPathFromOpenApiDocumentInApplyStyleForPowerShellOpenApiSty
             subsetOpenApiDocument = OpenApiService.ApplyStyle(OpenApiStyle.PowerShell, subsetOpenApiDocument);
 
             // Assert
-            Assert.Equal(4457, subsetOpenApiDocument.Paths.Count);
+            Assert.Equal(4544, subsetOpenApiDocument.Paths.Count);
```
Comment (ddyett) [inline OpenAPIService.Test/OpenAPIServiceShould.cs:300]: do we really need this assertion.  seems brittle.

File: UriMatchService/UriTemplateMatcher.cs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+// ------------------------------------------------------------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace UriMatchingService
```
Comment (ddyett) [inline UriMatchService/UriTemplateMatcher.cs:11]: can you add in code comments the library we got the code from and can you comments around the key changes that caused us to add this.

Comment (irvinesunday) [reply UriMatchService/UriTemplateMatcher.cs:22]: Resolved

File: OpenAPIService.Test/OpenAPIServiceShould.cs (hunk 1 above)
Comment (irvinesunday) [reply OpenAPIService.Test/OpenAPIServiceShould.cs:300]: We need to assert whether we generate all the paths to guard against an unprecedented change that breaks the path generation.

Comment (irvinesunday) [reply OpenAPIService.Test/OpenAPIServiceShould.cs:300]: This particular change was caused by the generation of Media Entity stream paths generation.

**Review by darrelmiller — DISMISSED** [review]

**Review by darrelmiller — APPROVED** [review]

**Review by ddyett — APPROVED** [review]

**Review by zengin — APPROVED** [review]

**Pull request merged**